pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/RangeIndexCreator.java

```diff
@@ -0,0 +1,506 @@
+
+  public RangeIndexCreator(File indexDir, FieldSpec fieldSpec, FieldSpec.DataType valueType, int numRanges,
+      int numDocsPerRange, int numDocs, int numValues)
+      throws IOException {
+    _valueType = valueType;
+    String columnName = fieldSpec.getName();
+    _rangeIndexFile = new File(indexDir, columnName + BITMAP_RANGE_INDEX_FILE_EXTENSION);
+    _tempValueBufferFile = new File(indexDir, columnName + VALUE_BUFFER_SUFFIX);
+    _tempDocIdBufferFile = new File(indexDir, columnName + DOC_ID_VALUE_BUFFER_SUFFIX);
+    _numValues = fieldSpec.isSingleValueField() ? numDocs : numValues;
+    int valueSize = valueType.size();
+    try {
+      //use DEFAULT_NUM_RANGES if numRanges is not set
+      if (numRanges < 0) {
+        numRanges = DEFAULT_NUM_RANGES;
+      }
+      if (numDocsPerRange < 0) {
+        _numDocsPerRange = (int) Math.ceil(_numValues / numRanges);
+      }
+
+      //Value buffer to store the actual values added
+      _tempValueBuffer = createTempBuffer((long) _numValues * valueSize, _tempValueBufferFile);
+
+      switch (_valueType) {
+        case INT:
+          _numberValueBuffer = new IntValueBuffer(_tempValueBuffer);
+          break;
+        case FLOAT:
+          _numberValueBuffer = new FloatValueBuffer(_tempValueBuffer);
+          break;
+        case LONG:
+          _numberValueBuffer = new LongValueBuffer(_tempValueBuffer);
+          break;
+        case DOUBLE:
+          _numberValueBuffer = new DoubleValueBuffer(_tempValueBuffer);
+          break;
+        default:
+          throw new UnsupportedOperationException("Range index is not supported for columns of data type:" + valueType);
+      }
+
+      //docId  Buffer
+      _docIdValueBuffer = createTempBuffer((long) _numValues * Integer.BYTES, _tempDocIdBufferFile);
+      _docIdBuffer = new IntValueBuffer(_docIdValueBuffer);
+    } catch (Exception e) {
+      destroyBuffer(_tempValueBuffer, _tempValueBufferFile);
+      destroyBuffer(_tempValueBuffer, _tempDocIdBufferFile);
+      throw e;
+    }
+  }
+
+  @Override
+  public void add(int dictId) {
+    _numberValueBuffer.put(_nextDocId, dictId);
+    _docIdBuffer.put(_nextDocId, _nextDocId);
+    _nextDocId = _nextDocId + 1;
+  }
+
+  @Override
+  public void add(int[] dictIds, int length) {
+    for (int i = 0; i < length; i++) {
+      int dictId = dictIds[i];
+      _numberValueBuffer.put(_nextValueId, dictId);
+      _docIdBuffer.put(_nextValueId, _nextDocId);
+      _nextValueId = _nextValueId + 1;
+    }
+    _nextDocId = _nextDocId + 1;
+  }
+
+  @Override
+  public void addDoc(Object document, int docIdCounter) {
+    throw new IllegalStateException("Bitmap inverted index creator does not support Object type currently");
+  }
+
+  @Override
+  public void seal()
+      throws IOException {
+    if (TRACE) {
+      LOGGER.info("Before sorting");
+      dump();
+    }
+
+    //sort the value buffer, change the docId buffer to maintain the mapping
```

Comment:
Let's add a diagram here showing the contents of _numberValueBuffer and _docIdBuffer before and after sorting to explain the process.